tests/MongoDB.Driver.Tests/Specifications/transactions/TransactionTestRunner.cs

```diff
@@ -117,6 +137,37 @@ private void Run(BsonDocument shared, BsonDocument test)
             }
         }
 
+        /// <summary>
+        /// Temporary patch until SERVER-39704 is resolved.
+        /// </summary>
+        private void PrimeShardRoutersWithDistinctCommand()
+        {
+            var connectionString = CoreTestConfiguration.ConnectionStringWithMultipleShardRouters.ToString();
+            var clientSettings = MongoClientSettings.FromUrl(new MongoUrl(connectionString));
+            var shardRouters = clientSettings.Servers.ToList();
+            MongoClientSettings CloneSettingsWithOneServer(MongoClientSettings settings, MongoServerAddress server)
+            {
+                var clone = settings.Clone();
+                clone.Server = server;
+                return clone;
+            }
+
+            var firstShardRouterClient = new MongoClient(CloneSettingsWithOneServer(clientSettings, shardRouters[0]));
+            var secondShardRouterClient = new MongoClient(CloneSettingsWithOneServer(clientSettings, shardRouters[1]));
+
+            void ExecuteDistinct(IMongoClient client)
+            {
+                client.GetDatabase(_databaseName)
+                    .GetCollection<BsonDocument>(_collectionName)
+                    .Distinct(
+                        new StringFieldDefinition<BsonDocument, BsonValue>("_id"), 
+                        FilterDefinition<BsonDocument>.Empty, 
+                        new DistinctOptions());
+            }
+            ExecuteDistinct(firstShardRouterClient);
+            ExecuteDistinct(secondShardRouterClient);
```

Comment:
This whole method is unnecessarily complicated with the local functions and hard coded assumptions about how many shard routers there might be. It can be as simple as this:

    var clientSettings = DriverTestConfiguration.Client.Settings;
    foreach (var server in clientSettings.Servers)
    {
        var singleServerSettings = clientSettings.Clone();
        singleServerSettings.Server = server;
        var singleServerClient = new MongoClient(singleServerSettings);
        var database = singleServerClient.GetDatabase(_databaseName);
        var collection = database.GetCollection<BsonDocument>(_collectionName);
        collection.Distinct<BsonValue>("_id", "{ }");
    }

Reply:
Since looping over singleServerClients might be useful in other scenarios it might be a good idea to add a helper method to DriverTestConfiguration to get a list of singleServerClients, then this method can be even simpler:

    foreach (var singleServerClient in DriverTestConfiguration.GetSingleServerClients())
    {
        var database = singleServerClient.GetDatabase(_databaseName);
        var collection = database.GetCollection<BsonDocument>(_collectionName);
        collection.Distinct<BsonValue>("_id", "{ }");
    }

Reply:
"Making it so" (i.e. refactoring) per our offline discussion 😸